Save session.id in session

JoeSouthan opened the pull request:
Fixes #44

I've refactored the sign/in out flow to use the session rather than a cookie.

Couple of things of note:
- This introduces an n+1 query, loading the Passwordless::Session, then the `authenticatable`
  - Could possibly store the Authenticatable in the session as well, but that adds complexity
  - Could employ some Rails caching to prevent loading the Session each time
- This is a breaking change in that `ControllerHelpers.passwordless_for` is now required to set up controllers with the correct resource to authenticate against, however I've made some of the other method params optional. Wondering if it's better to just remove the superfluous params?
- Instead of deprecating `ControllerHelpers#authenticate_by_cookie`, what do you think of removing it since it's a breaking change?

Let me know what you think!

mikker commented:
Nice!

- I don't think it's a problem to fetch both Session and Authenticatable. The user can lazy load the Authenticatable and, say, only look for session in their `require_user!` method, then only fetch the associated record if `current_user` is called. I don't think we should add the caching as it'll only add more complexity for a tiny gain.
- I don't really like the new `passwordless_for` in ApplicationController. Could we get the information we need from whatever the user has already put in their routes? I feel like if we have to change the apis it should be towards less _magic_ and more _normal_ Ruby/Rails code.
- What are the benefits of using session over cookie? Can we get the same, long expiry? Then I'm okay with switching. And if we do, you're right that `authenticate_by_cookie` isn't the best name 😁 

Thank you for doing this!

JoeSouthan commented:
Thanks for taking a look at the PR!

I've
- Refactored it to touch less of the existing API.
- Kept `authenticate_by_session` but removed the controller setup - I agree it was a bit too magic :tada: 

As it stood, cookies did expose implementation details (model type) and allowed for session fixation. I've added `reset_session` as suggested by the [Rails docs](https://guides.rubyonrails.org/security.html#session-fixation-countermeasures) to help mitigate this.

As to expiry, it will now honour whatever is in `Rails.application.config.session_store.expires_at` or `Session#expired?`.

mikker commented on test/controllers/deprecated_secrets_controller_test.rb:
```diff
@@ -0,0 +1,33 @@
+  end
+
+  def cookie_key(authenticatable_class)
+    DeprecatedSecretsController.new.send(:cookie_key, authenticatable_class)
```
Let's pull this method out into its own file so we can test it on its own and we don't need weird (in a Rails context) initialisation?

mikker commented on test/controllers/passwordless/sessions_controller_test.rb:
```diff
@@ -12,6 +12,10 @@ def create_session_for(user)
     end
 
+    def session_key(authenticatable_class)
+      Passwordless::SessionsController.new.send(:session_key, authenticatable_class)
```
Same as above.

mikker commented on app/controllers/passwordless/sessions_controller.rb:
```diff
@@ -104,8 +97,8 @@ def find_authenticatable
-    def find_session
-      Session.find_by!(
+    def passwordless_session
+      @passwordless_session ||= Session.find_by!(
```
Why do we memoize this if we only use it once?

mikker commented on app/models/passwordless/session.rb:
```diff
@@ -33,14 +31,18 @@ def timed_out?
       !!claimed_at
     end
 
+    def valid_session?
```
`session.valid_session?` has a double _session_. Let's rename to `session.valid?` ?

mikker commented on lib/passwordless/controller_helpers.rb:
```diff
@@ -3,6 +3,12 @@
+    # Returns the {Passwordless::Session} (if set) from the session.
+    # @return [Session, nil]
+    def current_passwordless_session(authenticatable_class)
+      @current_passwordless_session ||= Passwordless::Session.find_by(id: session[session_key(authenticatable_class)])
```
I don't really like to memoize on behalf of the user. My guess is it is only going to lead to weird, hard to figure out inconsistencies? _Rose memoization_ is widely enough known that we can assume people know to do it if they want it.

The readme already suggests to memoize `current_user` which is, probably, mostly what these methods will be used for.

mikker commented on lib/passwordless/controller_helpers.rb:
```diff
@@ -27,54 +34,98 @@ def build_passwordless_session(authenticatable)
+    # @see ModelHelpers#passwordless_with
+    def authenticate_by_session(authenticatable_class)
+      return unless current_passwordless_session(authenticatable_class)&.valid_session?
+      @current_authenticatable ||= current_passwordless_session(authenticatable_class).authenticatable
```
Same concern about memoization as above.

mikker commented on lib/passwordless/controller_helpers.rb:
```diff
@@ -27,54 +34,98 @@ def build_passwordless_session(authenticatable)
+        build_passwordless_session(record).tap { |s| s.save! }
+      end
+
+      sign_out(authenticatable_class)
```
Is this to make sure to clear the old cookies?

mikker commented on lib/passwordless/controller_helpers.rb:
```diff
@@ -27,54 +34,98 @@ def build_passwordless_session(authenticatable)
       cookies.delete(key)
+      # /deprecated
+
+      reset_session
```
Nice find in the docs 👏

mikker commented:
Thanks a bunch for yet another great addition 💙💚💛💜❤️ !

mikker commented:
@JoeSouthan I'm not seeing session persistence in my app using this. Sessions behave like session and reset between browser sessions, no matter what I set as `Rails.application.config.session_store.expires_after`.

mikker commented:
Actually, I think that might've been due to this and not a session vs cookies thing: https://github.com/mikker/passwordless/pull/61
